File: progs/seis/image/Kirch_3d_time.cpp


/*$


=head1 NAME

Kirch_3d_time - Kirchhoff 3D migration

=head1 SYNOPSIS

Kirch_3d_time  image=in.H data=out.H  velfile=tfile.H  pars

=cut

>*/
#include <tbb/blocked_range.h>
#include <tbb/parallel_for.h>
#include <tbb/tbb.h>

#include <float3DReg.h>
#include <float5DReg.h>
#include <seplib.h>
#include <memory>

#include <float3DReg.h>

#include <genericIO.h>

#include <thread>
#include "ioModes.h"
#include "kirchTime.h"

using namespace SEP;
using namespace SEP::KirchTime;
int main(int argc, char** argv) {
  // Setup IO

  ioModes modes = ioModes(argc, argv);

  std::shared_ptr<genericIO> ioIn = modes.getInputIO(),
                             ioOut = modes.getOutputIO();

  std::shared_ptr<paramObj> par = modes.getParamObj();

  // Amount of memory to use
  long long maxsize =
      (par->getDocInt("maxsize", "Maximum amount of memory to use in GBs", 30) -
       2) *
      (long long)1024 * (long long)1024 * (long long)1024;

  // Initalize files
  std::shared_ptr<vel3d> vel;

  try {
    vel.reset(new vel3d(ioIn->getDocRegFile(
        "velocity", "Velocity file (nt,cmpx,cmpy)", usageIn)));
  } catch (SEPException& e) {
    std::cout << e.what() << std::endl;
    exit(-1);
  }
  bool modelIt = par->getDocBool(
      "model", "Whether to model[true] or migrate[false]", false);

  std::shared_ptr<data5DReg> data, image5;
  std::shared_ptr<data3DReg> image3;
  std::shared_ptr<float1DReg> wave;

  if (!modelIt) {
    std::shared_ptr<genericRegFile> dataF = ioIn->getDocRegFile(
        "data", "Data file (t,ox,oy,cmpx,cmpy)", SEP::usageIn);

    data.reset(new data5DReg(dataF));
    std::vector<axis> axData = data->getHyper()->getAxes();
    axis axX, axY;
    axX.n = par->getDocInt("nx", "Number of CMPs in X for output", axData[3].n);
    axX.o = par->getDocFloat("ox", "First CMPs in X for output", axData[3].o);
    axX.d =
        par->getDocFloat("dx", "Sampling of CMPs in X for output", axData[3].d);
    axY.n = par->getDocInt("ny", "Number of CMPs in Y for output", axData[4].n);
    axY.o = par->getDocFloat("oy", "First  CMPs in y for output", axData[4].o);
    axY.d =
        par->getDocFloat("dy", "Sampling  CMPs in T for output", axData[4].d);
    image5.reset(new data5DReg(data, axX, axY));

    std::shared_ptr<genericRegFile> imageF = ioOut->getDocRegFile(
        "image",
        "Image file (t,cmpx,cmpy)<--modeling migrating-->(t,ox,oy,cmpx,cmpy)",
        SEP::usageOut);
    imageF->setHyper(image5->getHyper());
    image5->setFile(imageF);
  } else {
    std::shared_ptr<genericRegFile> imageF = ioIn->getDocRegFile(
        "image",
        "Image file (t,cmpx,cmpy)<--modeling migrating-->(t,ox,oy,cmpx,cmpy)",
        SEP::usageIn);

    image3.reset(new data3DReg(imageF));

    std::vector<axis> axImage = image3->getHyper()->getAxes();

    axis axX, axY;
    axis axOX, axOY;
    axX.n =
        par->getDocInt("nx", "Number of CMPs in X for output", axImage[1].n);

    axX.o = par->getDocFloat("ox", "First CMPs in X for output", axImage[1].o);
    axX.d = par->getDocFloat("dx", "Sampling of CMPs in X for output",
                             axImage[1].d);
    axY.n =
        par->getDocInt("ny", "Number of CMPs in Y for output", axImage[2].n);
    axY.o = par->getDocFloat("oy", "First CMPs in Y for output", axImage[2].o);
    axY.d = par->getDocFloat("dy", "Sampling of CMPs in Y for output",
                             axImage[2].d);
    axOX.n = par->getDocInt("nox", "Number of Offsets in X for output");
    axOX.o = par->getDocFloat("oox", "First Offsets in X for output");
    axOX.d = par->getDocFloat("dox", "Sampling of Offsets in X for output");
    axOY.n = par->getDocInt("noy", "Number of Offsets in Y for output");
    axOY.o = par->getDocFloat("ooy", "First Offsets in Y for output");
    axOY.d = par->getDocFloat("doy", "Sampling of Offsets in Y for output");
    std::shared_ptr<data5DReg> d(new data5DReg(image3, axOX, axOY, axX, axY));
    data = d;

    std::shared_ptr<genericRegFile> dataF = ioOut->getDocRegFile(
        "data", "Data file (t,ox,oy,cmpx,cmpy)", SEP::usageOut);

    dataF->setHyper(data->getHyper());
    d->setFile(dataF);

    std::shared_ptr<genericRegFile> waveF = ioIn->getDocRegFile(
        "wave", "Wavelet (only needed for modeling)", SEP::usageIn);
    std::shared_ptr<float1DReg> w(new float1DReg(waveF->getHyper()));
    wave = w;

    std::vector<int> fw(1, 0), jw(1, 1), nw(1, waveF->getHyper()->getAxis(1).n);
    try {
      waveF->readWindow(nw, fw, jw, wave);
    } catch (SEPException& x) {
      par->error(x.what());
    }
  }

  // Make sure velocity is at least as big as the output
  data->checkLogic(vel);

  if (!modelIt) {
    // Read velocity
    try {
      maxsize -= vel->readVelocity(image5->getHyper(), maxsize);
    } catch (SEPException& e) {
      std::cout << e.what() << std::endl;
      exit(-1);
    }
  } else {
    try {
      maxsize -= vel->readVelocity(image3->getHyper(), maxsize);

    } catch (SEPException& e) {
      std::cout << e.what() << std::endl;
      exit(-1);
    }
  }
  std::shared_ptr<kirchhoffTime> kir(
      new kirchhoffTime(vel, data, maxsize, par->getFloat("aper")));

  if (modelIt) {
    try {
      kir->modelData(image3, wave);

    } catch (SEPException& e) {
      std::cout << e.what() << std::endl;
      exit(-1);
    }

    data->getFile()->writeDescription();

  } else {
    try {
      kir->migrateData(image5);
    } catch (SEPException& e) {
      std::cout << e.what() << std::endl;
      exit(-1);
    }
    image5->getFile()->writeDescription();
  }
}

File: progs/seis/image/Kirch_3d_time.py
#!/usr/bin/env @Python3_EXECUTABLE@
import argparse
import genericIO
import KirchTime
import Hypercube


def migrateIt(args):
    """Function for Kirchhoff time migration"""
    inputIO = genericIO.io()
    outputIO = genericIO = io()
    if args.inputIO:
        inputIO = genericIO.io(args.inputIO)
    if args.outputIO:
        outputIO = genericIO.io(args.outputIO)
    vel3d = KirchTime.vel3d(inputIO.getRegFile("velocity"))
    data = KirchTime.data5DReg(file=inputIO.getRegFile(args.data))
    hyperD = data.getHyper()
    aX = hyperD.getAxis(4)
    aY = hyperD.getAxis(5)
    if args.nx:
        aX.n = args.nx
    if args.ox:
        aX.o = args.ox
    if args.dx:
        aX.d = args.dx
    if args.ny:
        aY.n = args.ny
    if args.oy:
        aY.o = args.oy
    if args.dy:
        aY.d = args.dy

    image = KirchTime.KirchTime.data5DReg(data=data, aX=aX, aY=aY)
    imageF = outputIO.getRegFile(
        args.image,
        fromHyper=image.getHyper(),
        storage="float",
        usage="output")
    image.setFile(imageF)
    maxsize = args.maxsize * 1024 * 1024 * 1024
    data.checkLogic(vel3d)
    maxsize -= vel3d.readVelocity(image.getHyper(), maxsize)
    kir = kirchTime.kirchTime(vel3d, data, maxsize, args.aper)
    kir.migrateData(image)
    image.getFile().close()


def modelIt(args):
    """Function for Kirchhoff time modeling"""
    inputIO = genericIO.io()
    outputIO = genericIO.io()
    if args.inputIO:
        inputIO = genericIO.io(args.inputIO)
    if args.outputIO:
        outputIO = genericIO.io(args.outputIO)
    vel3d = KirchTime.vel3d(inputIO.getRegFile("velocity"))
    imageF = inputIO.getRegFile(args.image)
    image = KirchTime.data3DReg(imageF)
    hyperI = image.getHyper()
    aX = hyperI.getAxis(2)
    aY = hyperI.getAxis(3)
    aOX = Hypercube.axis(n=1, d=.1)
    aOY = Hypercube.axis(n=1, d=.1)
    if args.nx:
        aX.n = args.nx
    if args.ox:
        aX.o = args.ox
    if args.dx:
        aX.d = args.dx
    if args.ny:
        aY.n = args.ny
    if args.oy:
        aY.o = args.oy
    if args.dy:
        aY.d = args.dy
    if args.nox:
        aOX.n = args.nox
    if args.oox:
        aOX.o = args.oox
    if args.dox:
        aOX.d = args.dox
    if args.noy:
        aOY.n = args.noy
    if args.ooy:
        aOY.o = args.ooy
    if args.doy:
        aOY.d = args.doy
    data = KirchTime.data5DReg(image=image, aOX=aOX, aOY=aOY, aX=aX, aY=aY)
    dataF = outputIO.getRegFile(
        args.data,
        fromHyper=data.getHyper(),
        storage="float",
        usage="output")
    data.setFile(dataF)
    print("w1")
    wave = inputIO.getVector(args.wavelet)
    print("w1")

    maxsize = args.maxsize * 1024 * 1024 * 1024
    print("w1")

    data.checkLogic(vel3d)
    print("w1")

    maxsize -= vel3d.readVelocity(image.getHyper(), maxsize)
    print("w1")

    kir = KirchTime.kirchTime(vel3d, data, maxsize, args.aper)
    print("w1")

    kir.modelData(image, wave)
    data.getFile().close()
parser = argparse.ArgumentParser(
    description="Kirchhoff 3D migration / modeling")
subParser = parser.add_subparsers(help="sub command help")
modes = genericIO.ioModes.getIOs()
defType = genericIO.ioModes.getDefaultType()


def jointArguments(parser):
    """Arguments for both options"""
    parser.add_argument(
        "-inputIO",
        type=str,
        choices=modes, default=defType,
        help="Input file types")
    parser.add_argument(
        "-outputIO",
        type=str,
        choices=modes, default=defType,
        help="Output file types")
    parser.add_argument(
        "-datapath",
        type=str,
        help="Location for binary (for some IOs)")
    parser.add_argument(
        "-nx",
        type=int,
        help="Number of cmps in x for output")
    parser.add_argument(
        "-ox",
        type=float,
        help="First  cmp x location for output")
    parser.add_argument(
        "-dx",
        type=float,
        help="Sampling of cmps in x for output")
    parser.add_argument(
        "-ny",
        type=int,
        help="Number of cmps in y for output")
    parser.add_argument(
        "-oy",
        type=float,
        help="First  cmp y location for output")
    parser.add_argument(
        "-dy",
        type=float,
        help="Sampling of cmps in y for output")
    parser.add_argument(
        "-maxsize",
        type=int,
        default=24,
        help="Amount of memory to use in GB")

modelParser = subParser.add_parser("model")
modelParser.add_argument(
    "velocity",
    type=str,
    help="VRMS (nt,ncmpx,ncmpy) [input]")
modelParser.add_argument(
    "wavelet",
    type=str,
    help="Wavelet to convovle data with")


modelParser.add_argument(
    "image",
    type=str,
    help="Image (nt,ncmpx,ncmpy) [input]")
modelParser.set_defaults(func=modelIt)
modelParser.add_argument(
    "data", type=str, help="data (nt,offx,offy,ncmpx,ncmpy [output]")
modelParser.add_argument("aper", type=float, help="Aperture ofr migration")


modelParser.add_argument(
    "-nox",
    type=int,
    help="Number of offset in x for output")
modelParser.add_argument(
    "-oox",
    type=float,
    help="First  offset x location for output")
modelParser.add_argument(
    "-dox",
    type=float,
    help="Sampling of offset in x for output")
modelParser.add_argument(
    "-noy",
    type=int,
    help="Number of offset in x for output")
modelParser.add_argument(
    "-ooy",
    type=float,
    help="First  offset y location for output")
modelParser.add_argument(
    "-doy",
    type=float,
    help="Sampling of offset in y for output")


migrateParser = subParser.add_parser("migrate")
migrateParser.add_argument("velocity", help="VRMS (nt,ncmpx,ncmpy) [input]")

migrateParser.add_argument(
    "data", type=str, help="Image (nt,offx,offy,ncmpx,ncmpy) [input]")
migrateParser.add_argument(
    "image", type=str, help="data (nt,offx,offy,ncmpx,ncmpy [output]")
migrateParser.add_argument("aper", type=float, help="Aperture ofr migration")
migrateParser.set_defaults(func=migrateIt)


migrateParser = jointArguments(migrateParser)
modelParser = jointArguments(modelParser)
args = parser.parse_args()


paramArgs = genericIO.argParseParams(vars(args))

args.func(args)
